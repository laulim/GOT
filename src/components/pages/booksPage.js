import React, {Component} from 'react';
import RowBlock from '../rowBlock';
import ItemList from '../itemList';
import ItemDetails, {Field} from '../itemDetails';
import ErrorMessage from '../errorMessage';
import GotService from '../../services/gotService';

export default class BooksPage extends Component {

  gotService = new GotService();
  state = {
    selectedBook: null,
    error: false
  }

  componentDidCatch() {
    this.setState({
      error: true
    })
  }

  onItemSelected = (id) => {
    this.setState({
      selectedBook: id
    })
  }

  render () {
    if (this.state.error) {
      return <ErrorMessage/>
    }
    const page = Math.floor(Math.random()*4 + 1);

    const itemList = (
      <ItemList 
        onItemSelected={this.onItemSelected}
        getData={this.gotService.getAllBooks}
        pageDetails={[page, 3]}
        renderItem={({name}) => name}
      />
    )

    const itemDetails = (
      <ItemDetails 
        itemId={this.state.selectedBook}
        getData={this.gotService.getBook}
        emptyTitle='book'
      >
        <Field field='numberOfPages' label='Number of pages' />
        <Field field='publisher' label='Publisher' />
        <Field field='released' label='Released' />
      </ItemDetails>
    )

    return (
      <RowBlock left={itemList} right={itemDetails}/>
    )
  }
}