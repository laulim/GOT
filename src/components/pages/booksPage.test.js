import React from 'react';
import {shallow} from 'enzyme';
import {BooksPage} from './booksPage';

describe('Testing <BooksPage/>', () => {
  const booksPage = shallow(<BooksPage/>);

  it('BooksPage has been rendered', () => {
    expect(booksPage).toMatchSnapshot();
  });

  it('BooksPage state "error" is false', () => {
    expect(booksPage.state().error).toBeFalsy();
  });

  it('Testing componentDidCatch, expect state "error" is true', () => {
    booksPage.instance().componentDidCatch();
    expect(booksPage.state().error).toBeTruthy();
  })
})