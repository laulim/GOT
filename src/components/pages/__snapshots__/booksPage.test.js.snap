// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <BooksPage/> BooksPage has been rendered 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BooksPage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <RandomItem
          btnName="book"
          fieldsList={
            Array [
              Array [
                "numberOfPages",
                "Number of pages",
              ],
              Array [
                "publisher",
                "Publisher",
              ],
              Array [
                "released",
                "Released",
              ],
            ]
          }
          getData={[Function]}
          itemIdInterval={
            Array [
              10,
              1,
            ]
          }
        />,
        <ItemList
          getData={[Function]}
          onItemSelected={[Function]}
          pageDetails={
            Array [
              1,
              5,
            ]
          }
          renderItem={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "btnName": "book",
          "fieldsList": Array [
            Array [
              "numberOfPages",
              "Number of pages",
            ],
            Array [
              "publisher",
              "Publisher",
            ],
            Array [
              "released",
              "Released",
            ],
          ],
          "getData": [Function],
          "itemIdInterval": Array [
            10,
            1,
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "getData": [Function],
          "onItemSelected": [Function],
          "pageDetails": Array [
            1,
            5,
          ],
          "renderItem": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <RandomItem
            btnName="book"
            fieldsList={
              Array [
                Array [
                  "numberOfPages",
                  "Number of pages",
                ],
                Array [
                  "publisher",
                  "Publisher",
                ],
                Array [
                  "released",
                  "Released",
                ],
              ]
            }
            getData={[Function]}
            itemIdInterval={
              Array [
                10,
                1,
              ]
            }
          />,
          <ItemList
            getData={[Function]}
            onItemSelected={[Function]}
            pageDetails={
              Array [
                1,
                5,
              ]
            }
            renderItem={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "btnName": "book",
            "fieldsList": Array [
              Array [
                "numberOfPages",
                "Number of pages",
              ],
              Array [
                "publisher",
                "Publisher",
              ],
              Array [
                "released",
                "Released",
              ],
            ],
            "getData": [Function],
            "itemIdInterval": Array [
              10,
              1,
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "getData": [Function],
            "onItemSelected": [Function],
            "pageDetails": Array [
              1,
              5,
            ],
            "renderItem": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
